### PR TITLE
Full sync with logging for very slow state block commits

### LINKED ISSUE #4932: Reduce the frequency of Zebra performance logs
## Motivation

As part of #4747, we added some info-level performance logs to Zebra.

But performance is good enough now, so we can reduce the frequency of those logs.

## Tasks

We could downgrade all the logs to debug level, but it might still be useful to log them for very long tasks.

For example:
- the state can take up to a minute to commit a block, so we want to log that at trace or debug level
- rebuilding non-finalized note commitment trees can take 1-2 minutes, so we could log that at info level
- blocks timeout after 10 minutes, so we might want to log a warning before 10 minutes

The performance logging is implemented in:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/diagnostic.rs

### LINKED ISSUE #4937: Tracking: Commit blocks to state using a separate task
## Motivation

Zebra takes 10-15 minutes to commit some blocks to the state while checkpointing, around blocks 1,718,00 to 1,772,000.

The slow blocks are different on different runs.

This is unacceptable performance, because:
* it's much slower than `zcashd`
* Zebra will appear to hang for 15 minutes, which is a usability and security issue
* it causes warnings in the Zebra logs
* if it's remotely triggerable, it could be a denial of service risk

## Diagnosis

Zebra queues up to 1200 blocks, then commits them all in the same state request, after the missing block arrives. This can take up to 10 seconds per block.

### Design

Add a block commit task to the state, which runs in a separate thread. The task should be between the block queue and the block verifier.

We'll need to move the shared mutable chain state into the block commit task, so we will also need to redirect `StateService` read requests to the concurrent `ReadStateService`.

Here is a diagram of the new state design:
https://docs.google.com/drawings/d/1FXpAUlenDAjl8nkftrypdAPsj0jr-Ut9gZlSP57nuyc/edit

#### Stop Accessing Mutable Chain State

- [x] #5102

#### Setup Block Commit task

- [x] Add a new block commit task with unused channels

#### Add channels to send blocks to the task

- [x] Add a channel that handles finalized state `CommitFinalizedBlock` requests
- [x] Add a channel that handles non-finalized state `CommitBlock` requests
- [x] We want two channels so we can wait for the last finalized block before committing the first non-finalized block (by height)
    - The current implementation of this has a bug: #5125

#### Error Handling & Testing

- [x] Handle panics in the block commit task by panicking in the service
- [x] Testing - what new tests do we want
    - [x] #5315
    - [x] #2654

Optional tasks:

<details>

#### Optional Cleanup Tasks

Bug fixes:
- [x] #5126

Refactors:
- Make pending_utxos.respond() async using a channel, so we can use `ReadRequest::ChainUtxo` in `AwaitUtxo`

Renames & Formatting:
- Rename every instance of address `*` or `transparent_*` to `address_*`
- Put the `Request` and `Response` enums in a consistent order

</details>

## In Scope

- Non-finalized state
- Finalized state
- Running the task in a separate thread

## Out of Scope

We don't think we'll need to make these changes as part of this change:

<details>

- [x] #5101
- [x] #5103
  (this reduces the number of state requests from the syncer)

</details>

These are definitely out of scope:
- Other state refactors
- Other performance improvements
- Note commitment tree performance improvements

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We want to find out which parts of state block commits are slow in CI, because we can't reproduce them locally.

These are diagnostics for ticket #4937, where some blocks take 15 minutes to write to the state.

## Solution

- change info threshold to 5 minutes and warn to 9 minutes
- add slow code logging for block commits, block writes, and note commitment tree updates

## Review

This PR doesn't need review.

### Reviewer Checklist

  - [ ] Logging works

## Follow Up Work

Read the logs to work out why #4937 is happening

### CODE REVIEW

File: zebra-state/src/service/tests.rs [hunk 1]
```diff
@@ -314,12 +314,20 @@ proptest! {
...
+            .expect("chain contains valid blocks");
+
         let response = crate::service::check::legacy_chain(nu_activation_height, chain.into_iter().rev(), network)
             .map_err(|error| error.to_string());
 
         prop_assert_eq!(
             response,
-            Err("giving up after checking too many blocks".into())
+            Err(format!(
+                "could not find any transactions in recent blocks: checked 1000 blocks back from {tip_height:?}",
```
Comment (arya2) [inline zebra-state/src/service/tests.rs]: 1000 could be replaced with {MAX_LEGACY_CHAIN_BLOCKS} if we want this test to succeed when that's changed

**Review by arya2 — DISMISSED** [review]
Looks good to me

File: zebra-state/src/service/tests.rs (hunk 1 above)
Comment (teor2345) [reply zebra-state/src/service/tests.rs]: I accidentally pushed two PRs to this branch, this is actually from PR #4804. I'll sort it out tomorrow.

Comment (teor2345) [reply zebra-state/src/service/tests.rs]: Fixed in c334ba465 in PR #4804

**Review by oxarbitrage — COMMENTED** [review]
The code looks good for sure but i dont think we should be merging this things into the main branch.

This seems more suitable for a local test or for a specific branch outside main.

**Comment by teor2345** [comment]
> The code looks good for sure but i dont think we should be merging this things into the main branch.
> 
> This seems more suitable for a local test or for a specific branch outside main.

Unfortunately, I can't reproduce this issue on my local machine, so I need to add this debugging to the `main` branch.
Once we know what the issue is, we can remove it.

**Comment by oxarbitrage** [comment]
Sorry, still dont understand why you need to add it to the main branch. Can't you run the CI without merging ?

**Comment by teor2345** [comment]
> Sorry, still dont understand why you need to add it to the main branch. Can't you run the CI without merging ?

I'm not sure if it will work, but I can try.

I'll open a new PR to disable the logs for #4932, because that needs to go into `main`. Then I can make this PR and draft and do a full sync on it.

**Comment by teor2345** [comment]
I'm running a full sync here with these extra logs:
https://github.com/ZcashFoundation/zebra/actions/runs/2953663804

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
Let's try that full sync again, after the concurrency fixes in PR #5000:
https://github.com/ZcashFoundation/zebra/actions/runs/2959955933

**Comment by teor2345** [comment]
The logs I added didn't actually show any extra information, which is a bit of a surprise. It's possible we're just committing a large number of slow blocks at a time.

I added some more logging and started another full sync:
https://github.com/ZcashFoundation/zebra/actions/runs/2968581615

**Comment by teor2345** [comment]
It appears that committing blocks hangs because we sometimes commit over 1000 blocks at the same time, then return the result to the state caller (checkpointer or non-finalized state block committer):
> 2022-09-01T22:50:25.650944Z  WARN state: run time: very long queue_and_commit_finalized: 1774457..=1775638, 1181 blocks time=11m 24s module=zebra_state::service::finalized_state line=173
> 2022-09-01T22:50:25.650981Z  WARN state: run time: very long CommitFinalizedBlock time=11m 24s module=zebra_state::service line=697

https://github.com/ZcashFoundation/zebra/runs/8147047219?check_suite_focus=true#step:6:165

I'll update ticket #4937 with the details.

**Comment by teor2345** [comment]
This PR has achieved its goals, but we'll need the extra logging to test any fixes.

**Comment by teor2345** [comment]
This PR was only used for testing slow block commits, we can re-open it if we need to re-test.

**Pull request closed without merging**